File: examples/clj/src/com/rpl/agent/basic/mirror_agent.clj
(ns com.rpl.agent.basic.mirror-agent
  "Demonstrates cross-module agent invocation using mirror agents.

  Features demonstrated:
  - defagentmodule: Define an agent module
  - agentmodule: Define parameterized agent module
  - declare-cluster-agent: Create mirror reference to agent in another module
  - agent-client: Get client for mirror agent within agent node
  - agent-invoke: Invoke mirror agent across modules
  - rama/get-module-name: Get module name for cross-module references
  - Multiple module deployment to IPC"
  (:require
   [com.rpl.agent-o-rama :as aor]
   [com.rpl.rama :as rama]
   [com.rpl.rama.test :as rtest]))

;;; Module 1: Greeter agent that creates greeting messages
(aor/defagentmodule GreeterModule
  [topology]

  (-> topology
      (aor/new-agent "Greeter")
      (aor/node
       "greet"
       nil
       (fn [agent-node name]
         (let [greeting (str "Hello, " name "!")]
           (aor/result! agent-node greeting))))))

(defn create-mirror-module
  "Create Module 2: Mirror agent that invokes Greeter from Module 1"
  [greeter-module-name]
  (aor/agentmodule
   [topology]

   ;; Declare mirror reference to Greeter agent in GreeterModule
   (aor/declare-cluster-agent
    topology
    "GreeterMirror"
    greeter-module-name
    "Greeter")

   (-> topology
       (aor/new-agent "MirrorAgent")
       (aor/node
        "process"
        nil
        (fn [agent-node name]
          ;; Get client for the mirror agent
          (let [greeter-client (aor/agent-client agent-node "GreeterMirror")
                ;; Invoke the mirror agent (cross-module call)
                greeting       (aor/agent-invoke greeter-client name)
                ;; Add prefix to result
                result         (str "Mirror says: " greeting)]
            (aor/result! agent-node result)))))))

(defn -main
  "Run the mirror agent example demonstrating cross-module invocation"
  [& _args]
  (with-open [ipc (rtest/create-ipc)]
    ;; Launch GreeterModule first
    (rtest/launch-module! ipc GreeterModule {:tasks 1 :threads 1})
    (let [greeter-module-name (rama/get-module-name GreeterModule)
          ;; Create MirrorModule with reference to GreeterModule
          mirror-module       (create-mirror-module greeter-module-name)]

      ;; Launch MirrorModule
      (rtest/launch-module! ipc mirror-module {:tasks 1 :threads 1})

      ;; Get agent manager for MirrorModule
      (let [manager      (aor/agent-manager ipc (rama/get-module-name mirror-module))
            mirror-agent (aor/agent-client manager "MirrorAgent")]

        (println "Mirror Agent Results:")
        (println "Input: \"Alice\" -> Result:" (aor/agent-invoke mirror-agent "Alice"))
        (println "Input: \"Bob\" -> Result:" (aor/agent-invoke mirror-agent "Bob"))))))

File: examples/java/basic/src/main/java/com/rpl/agent/basic/MirrorAgent.java
package com.rpl.agent.basic;

import com.rpl.agentorama.AgentClient;
import com.rpl.agentorama.AgentManager;
import com.rpl.agentorama.AgentNode;
import com.rpl.agentorama.AgentTopology;
import com.rpl.agentorama.AgentsModule;
import com.rpl.agentorama.ops.RamaVoidFunction2;
import com.rpl.rama.test.InProcessCluster;
import com.rpl.rama.test.LaunchConfig;

/**
 * Java example demonstrating cross-module agent invocation using mirror agents.
 *
 * <p>This example demonstrates:
 *
 * <ul>
 *   <li>Multiple agent module definitions
 *   <li>declareClusterAgent: Create mirror reference to agent in another module
 *   <li>getAgentClient: Get client for mirror agent within agent node
 *   <li>invoke: Invoke mirror agent across modules
 *   <li>getModuleName: Get module name for cross-module references
 *   <li>Multiple module deployment to IPC
 * </ul>
 */
public class MirrorAgent {

  /**
   * Module 1: Greeter agent that creates greeting messages.
   *
   * <p>This module contains a simple agent that takes a name and returns a greeting.
   */
  public static class GreeterModule extends AgentsModule {

    @Override
    protected void defineAgents(AgentTopology topology) {
      topology.newAgent("Greeter").node("greet", null, new GreetFunction());
    }
  }

  /** Node function that creates a greeting message. */
  public static class GreetFunction implements RamaVoidFunction2<AgentNode, String> {

    @Override
    public void invoke(AgentNode agentNode, String name) {
      String greeting = "Hello, " + name + "!";
      agentNode.result(greeting);
    }
  }

  /**
   * Module 2: Mirror agent that invokes Greeter from Module 1.
   *
   * <p>This module declares a mirror reference to the Greeter agent and uses it to demonstrate
   * cross-module agent invocation.
   */
  public static class MirrorModule extends AgentsModule {
    private final String greeterModuleName;

    public MirrorModule(String greeterModuleName) {
      this.greeterModuleName = greeterModuleName;
    }

    @Override
    protected void defineAgents(AgentTopology topology) {
      // Declare mirror reference to Greeter agent in GreeterModule
      topology.declareClusterAgent("GreeterMirror", greeterModuleName, "Greeter");

      topology.newAgent("MirrorAgent").node("process", null, new ProcessFunction());
    }
  }

  /** Node function that invokes the mirror agent and adds a prefix. */
  public static class ProcessFunction implements RamaVoidFunction2<AgentNode, String> {

    @Override
    public void invoke(AgentNode agentNode, String name) {
      // Get client for the mirror agent
      AgentClient greeterClient = agentNode.getAgentClient("GreeterMirror");

      // Invoke the mirror agent (cross-module call)
      String greeting = (String) greeterClient.invoke(name);

      // Add prefix to result
      String result = "Mirror says: " + greeting;
      agentNode.result(result);
    }
  }

  public static void main(String[] args) throws Exception {
    System.out.println("Starting Mirror Agent Example...");

    try (InProcessCluster ipc = InProcessCluster.create()) {
      // Launch GreeterModule first
      GreeterModule greeterModule = new GreeterModule();
      ipc.launchModule(greeterModule, new LaunchConfig(1, 1));
      String greeterModuleName = greeterModule.getModuleName();

      // Launch MirrorModule with reference to GreeterModule
      MirrorModule mirrorModule = new MirrorModule(greeterModuleName);
      ipc.launchModule(mirrorModule, new LaunchConfig(1, 1));

      // Get agent manager for MirrorModule
      String mirrorModuleName = mirrorModule.getModuleName();
      AgentManager manager = AgentManager.create(ipc, mirrorModuleName);
      AgentClient mirrorAgent = manager.getAgentClient("MirrorAgent");

      // Invoke the mirror agent
      System.out.println("\nMirror Agent Results:");
      System.out.println("Input: \"Alice\" -> Result: " + mirrorAgent.invoke("Alice"));
      System.out.println("Input: \"Bob\" -> Result: " + mirrorAgent.invoke("Bob"));
    }
  }
}
